### ISSUE TITLE
pure javascript playlist configuration interface

### ISSUE THREAD

**Issue by foliovision** [foliovision/freedom-video-player#63]
Original issue link: [#65 pure javascript playlist configuration interface](https://github.com/flowplayer/flowplayer/issues/65)

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Oct 26, 2012

As far as I can see fallback to rtmp cannot be handled automatically with html http configured playlist items.

See also #64.

imho flexible playlist handling requires javascript at some point. Experience: I had to refactor the fp3 playlist plugin to use an internal playlist even with the 'manual' setup to iron out some long-standing bugs, but still the internal setup is more flexible.

Should be part of the capability to install player into an empty container via JS only.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Oct 26, 2012

See also #63.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Feb 4, 2013

http://flowplayer.org/forum/#!/moot/6643

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Feb 4, 2013

http://flowplayer.org/forum/#!/moot/7259

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Feb 18, 2013

Our own playlist demos - e.g. http://flowplayer.org/standalone/playlist/basic.html et al. - are blissfully unaware of the fact that playlists cannot work over rtmp: They configure rtmp, plus an rtmp address in the video tag, and therefore all fail in Flash: http://flowplayer.org/standalone/playlist/basic.html?engine=flash plays of course only the first item and then errors out.
Elevating to critical to raise awareness.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Feb 18, 2013

http://example.flowplayer.org/bug/65.html

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/265438?v=4' /> [nnarhinen](https://github.com/nnarhinen) commented on Feb 18, 2013

how would you like the interface to look like?

```
$('.flowplayer').flowplayer().playlist([{...},{...},{...}]);
```

?

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Feb 18, 2013

Yes - however, it would be inconsistent with the _array_ argument for load/play - I never understood why it had to be a list of single-member objects instead of just an object with multiple type-src mappings, i.e.

```
[{mp4: "video.mp4"}, {webm: "video.webm"}, {ogg: "video.ogv"}]
```

instead of simply:

```
{mp4: "video.mp4", webm: "video.webm", ogg: "video.ogv"}
```

Because braces are cewl? ;-)  - The latter would be easier to pick sources from, or not?
The interface however, is the easiest part imho - so far my meagre attempts have been thwarted by the fact that everything is so intertwined with the html, like next()/prev() triggering a simulated click() on those elements, and the click in turn triggers playback ...

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Feb 18, 2013

@nnarhinen - this ominous click imho is the reason that https://github.com/flowplayer/flowplayer/issues/58 is still not solved with non-splash setups in browsers which play mp4 (the anchors are mp4): the last clip is omitted.
I will make a demo on example.flowplayer.org for that.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Feb 18, 2013

@nnarhinen - make that #181 (migrated).

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/265438?v=4' /> [nnarhinen](https://github.com/nnarhinen) commented on Feb 18, 2013

@blacktrash some experimental support can now be found from http://example.flowplayer.org/not-indexed/javascript-playlist.html

This playlist thing is such a elephant I chose to take it one bite at a time.

Now you can do (as you can see from the example)

``` javascript
$(function() {
      var flowplayer = $('.flowplayer').flowplayer();
      flowplayer.conf.playlist = [
            [
               {'video/webm': 'http://stream.flowplayer.org/night3/640x360.webm'},
               {'video/mp4': 'http://stream.flowplayer.org/night3/640x360.mp4'},
               {'video/ogg': 'http://stream.flowplayer.org/night3/640x360.ogv'}
            ],
            [
               {'video/webm': 'http://stream.flowplayer.org/night5/640x360.webm'},
               {'video/mp4': 'http://stream.flowplayer.org/night5/640x360.mp4'},
               {'video/ogg': 'http://stream.flowplayer.org/night5/640x360.ogv'}
            ],
            [
               {'video/webm': 'http://stream.flowplayer.org/night6/640x360.webm'},
               {'video/mp4': 'http://stream.flowplayer.org/night6/640x360.mp4'},
               {'video/ogg': 'http://stream.flowplayer.org/night6/640x360.ogv'}
            ],
            [
               {'video/webm': 'http://stream.flowplayer.org/night1/640x360.webm'},
               {'video/mp4': 'http://stream.flowplayer.org/night1/640x360.mp4'},
               {'video/ogg': 'http://stream.flowplayer.org/night1/640x360.ogv'}
            ]
         ];
       });
```

To make a playlist which will respond to flowplayer.next(), flowplayer.prev() (and also automatically to .fp-next and .fp-prev clicks.

The compiled javascript is in http://example.flowplayer.org/flowplayer-javascript-playlist.js if you want to test some setups.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/265438?v=4' /> [nnarhinen](https://github.com/nnarhinen) commented on Feb 18, 2013

Ok so now when I have this initial sample working and it seems to now address the issue in #181 I ask if you @blacktrash could be so kind and try this out and tell me what should be improved. It isn't completely clear to me what is the desired amount of support and how much of it already is in there..

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Feb 18, 2013

@nnarhinen - thx, I'll look 2morrow, my 13 hour shift is over ;-)

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/265438?v=4' /> [nnarhinen](https://github.com/nnarhinen) commented on Feb 18, 2013

Same here, gonna hit the bed :)

2013/2/18 Christian Ebert notifications@github.com

> @nnarhinen https://github.com/nnarhinen - thx, I'll look 2morrow, my 13
> hour shift is over ;-)
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/flowplayer/flowplayer/issues/65#issuecomment-13744390.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Feb 22, 2013

@nnarhinen - as I'm working on #259 , we probably want a clip.cuepoints property, which also then triggers the creation of cuepoint elements when `generate_cuepoints == true`.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/265438?v=4' /> [nnarhinen](https://github.com/nnarhinen) commented on Feb 26, 2013

Yeah, so, now the playlist anchors get generated as we discussed in #181 

This initial implementation hardcodes the generated container to .fp-playlist and adds anchors directly there.

If preferred I can add configuration options for the container but I see it somwhat as a corner case, where someone wants to configure the playlist by hand and wants to have custom structure for the anchors.

This should be visible allready in http://example.flowplayer.org/bug/181.html

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/265438?v=4' /> [nnarhinen](https://github.com/nnarhinen) commented on Mar 15, 2013

This would need some more examples / tests, especially for engine switching by video type etc..

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Mar 16, 2013

@nnarhinen - when I'm back from holiday ;-)
We should enable rtmp for some of the playlist examples on the main page and use the JS interface there.
I am also now unclear regarding syntax:
So far load() took these as types: mp4, webm, ogg, flash, mpegurl. Is now both "video/mp4" and "mp4" allowed?

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/265438?v=4' /> [nnarhinen](https://github.com/nnarhinen) commented on Apr 2, 2013

Basically and in theory playlist acccepts whatever load() accepts.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/265438?v=4' /> [nnarhinen](https://github.com/nnarhinen) commented on Apr 13, 2013

Available in 5.4
